code/test-utils/src/test/java/com/adobe/marketing/tester/util/JSONAssertsParameterizedTests.kt

```diff
@@ -0,0 +1,416 @@
+        fun `should not fail because of alternate path`() {
+            assertExactMatch(expected, actual, typeMatchPaths = listOf(keypath))
+            assertTypeMatch(expected, actual, exactMatchPaths = listOf(keypath))
+        }
+    }
+
+    @RunWith(Parameterized::class)
+    class AlternatePathTypeTest(private val keypath: String, expectedJSONString: String, actualJSONString: String) {
+        private inline fun <T> tryOrNull(block: () -> T): T? {
+            return try {
+                block()
+            } catch (e: Throwable) {
+                null
+            }
+        }
+
+        private val expected: Any = (tryOrNull { JSONObject(expectedJSONString) } ?: tryOrNull { JSONArray(expectedJSONString) })!!
+        private val actual: Any? = tryOrNull { JSONObject(actualJSONString) } ?: tryOrNull { JSONArray(actualJSONString) }
+
+        companion object {
+            private fun createParameterCase(key: String, pair: Pair<Any, Any>, transform: (Any) -> String): Array<String> {
+                return arrayOf(key, transform(pair.first), transform(pair.second))
+            }
+
+            private val values = listOf(
+                Pair(1, 2),
+                Pair("a", "b"),
+                Pair(1.0, 2.0),
+                Pair(true, false),
+            )
+
+            @JvmStatic
+            @Parameterized.Parameters(name = "{index}: test with key={0}, expected={1}, actual={2}")
+            fun data(): Collection<Array<String>> {
+                return listOf(
+                    *values.map { pair -> createParameterCase("key1", pair) { """{ "key1": $it }""" } }.toTypedArray(),
+                    *values.map { pair -> createParameterCase("[0]", pair) { """[$it]""" } }.toTypedArray(),
+                    *values.map { pair -> createParameterCase("[*]", pair) { """[$it]""" } }.toTypedArray(),
+                )
+            }
+        }
+
+        @Test
+        fun `should apply alternate path to matching logic`() {
+            assertExactMatch(expected, actual, typeMatchPaths = listOf(keypath))
+            Assert.assertThrows(AssertionError::class.java) {
+                assertTypeMatch(expected, actual, exactMatchPaths = listOf(keypath))
+            }
+        }
+    }
+
+    @RunWith(Parameterized::class)
+    class SpecialKeyAlternatePathTest(private val keypath: String, expectedJSONString: String, actualJSONString: String) {
+        private val expected: JSONObject = JSONObject(expectedJSONString)
+        private val actual: JSONObject = JSONObject(actualJSONString)
+
+        companion object {
+            private inline fun <reified R> IntRange.mapToArray(transform: (Int) -> R): Array<R> {
+                return Array(this.last - this.first + 1) { index -> transform(this.first + index) }
+            }
+
+            private fun createParameterCase(key: String, transform: (Int) -> String): Array<String> {
+                return arrayOf(key, *(1..2).mapToArray(transform))
+            }
```

Comment:
this one became quite hard to read, would it be worth it to keep this in the old expanded format?

Reply:
thanks for the updates, the new format with testCase looks much better